http/events.go

```diff
@@ -0,0 +1,131 @@
+package http
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/vault"
+	"google.golang.org/protobuf/encoding/protojson"
+	"nhooyr.io/websocket"
+)
+
+var eventTypeRegex = regexp.MustCompile(`.*/events/subscribe/(.*)`)
+
+// handleEventsSubscribeWebsocket runs forever, returning a websocket error code and reason
+// only if the connection closes or there was an error.
+func handleEventsSubscribeWebsocket(ctx context.Context, core *vault.Core, ns *namespace.Namespace, eventType logical.EventType, conn *websocket.Conn, remoteAddr string, json bool) (websocket.StatusCode, string, error) {
+	events := core.Events()
+	closer, ch, err := events.Subscribe(ctx, ns, eventType)
+	if err != nil {
+		core.Logger().Info("Error subscribing", "error", err)
+		return websocket.StatusUnsupportedData, "Error subscribing", nil
+	}
+	defer closer.Close()
+
+	for {
+		select {
+		case <-ctx.Done():
+			core.Logger().Info("Websocket context is done, closing the connection")
+			return websocket.StatusNormalClosure, "", nil
+		case message := <-ch:
+			core.Logger().Debug("Sending message to websocket", "message", message, "remoteAddr", remoteAddr)
```

Comment:
IP addresses can be considered PII, so I typically try to avoid logging them unless really necessary, and the IP address also isn't enough to uniquely identify a stream of events. Is the idea here to help correlate between clients and server log messages? Maybe it would be useful to generate a kind of request ID (UUID-style), except more like a connection/session ID, that gets added to each event we send back for a single stream? Between that and event IDs, we should be able to correlate at whatever level is needed.

That doesn't have to be in this PR, just thinking aloud really.

Reply:
Oh good call. And yeah, I thought initially about making it easier to figure out what websockets were active to who. (The ServerAddr also includes the port, so it was helping some.) But, yeah, I'll remove the IP address for now.